// @google-cloud/vertexai/build/src/functions/post_request.js
"use strict";
/**
 * @license
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.postRequest = void 0;
const API_BASE_PATH = 'aiplatform.googleapis.com';
const errors_1 = require("../types/errors");
const constants = require("../util/constants");
/**
 * Makes a POST request to a Vertex service
 * @ignore
 */
async function postRequest({ region, project, resourcePath, resourceMethod, token, data, apiEndpoint, requestOptions, apiVersion = 'v1', }) {
    const vertexBaseEndpoint = apiEndpoint !== null && apiEndpoint !== void 0 ? apiEndpoint : `${region}-${API_BASE_PATH}`;
    let vertexEndpoint = `https://${vertexBaseEndpoint}/${apiVersion}/projects/${project}/locations/${region}/${resourcePath}:${resourceMethod}`;
    // Use server sent events for streamGenerateContent
    if (resourceMethod === constants.STREAMING_GENERATE_CONTENT_METHOD) {
        vertexEndpoint += '?alt=sse';
    }
    if ((requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiClient) &&
        ((requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiClient.includes('\n')) ||
            (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiClient.includes('\r')))) {
        throw new errors_1.ClientError('Found line break in apiClient request option field, please remove ' +
            'the line break and try again.');
    }
    const extraHeaders = (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiClient)
        ? { 'X-Goog-Api-Client': requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.apiClient }
        : {};
    return fetch(vertexEndpoint, {
        ...getFetchOptions(requestOptions),
        method: 'POST',
        headers: {
            Authorization: `Bearer ${token}`,
            'Content-Type': 'application/json',
            'User-Agent': constants.USER_AGENT,
            ...extraHeaders,
        },
        body: JSON.stringify(data),
    });
}
exports.postRequest = postRequest;
function getFetchOptions(requestOptions) {
    const fetchOptions = {};
    if (!requestOptions ||
        requestOptions.timeout === undefined ||
        requestOptions.timeout < 0) {
        return fetchOptions;
    }
    const abortController = new AbortController();
    const signal = abortController.signal;
    setTimeout(() => abortController.abort(), requestOptions.timeout);
    fetchOptions.signal = signal;
    return fetchOptions;
}
